peer/abstractlist/list.go

```diff
@@ -0,0 +1,620 @@
+// Updates may be interleaved with Start and Stop in any order any number of
+// times.
+func (pl *List) Update(updates peer.ListUpdates) error {
+	if len(updates.Additions) == 0 && len(updates.Removals) == 0 {
+		return nil
+	}
+
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	if !pl.once.IsRunning() {
+		return pl.updateOffline(updates)
+	}
+	return pl.updateOnline(updates)
+}
+
+// updateOnline must be run under a list lock.
+func (pl *List) updateOnline(updates peer.ListUpdates) error {
+	var errs error
+	for _, id := range updates.Removals {
+		errs = multierr.Append(errs, pl.remove(id))
+	}
+
+	add := updates.Additions
+	if !pl.noShuffle {
+		add = shuffle(pl.randSrc, add)
+	}
+
+	for _, id := range add {
+		errs = multierr.Append(errs, pl.add(id))
+	}
+	return errs
+}
+
+// updateOffline must be run under a list lock.
+func (pl *List) updateOffline(updates peer.ListUpdates) error {
+	var errs error
+	for _, id := range updates.Removals {
+		errs = multierr.Append(errs, pl.removeOffline(id))
+	}
+	for _, id := range updates.Additions {
+		errs = multierr.Append(errs, pl.addOffline(id))
+	}
+	return errs
+}
+
+// Start notifies the List that requests will start coming
+func (pl *List) Start() error {
+	return pl.once.Start(pl.start)
+}
+
+func (pl *List) start() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.offlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// Stop notifies the List that requests will stop coming
+func (pl *List) Stop() error {
+	return pl.once.Stop(pl.stop)
+}
+
+func (pl *List) stop() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.onlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// IsRunning returns whether the peer list is running.
+func (pl *List) IsRunning() bool {
+	return pl.once.IsRunning()
+}
+
+// add retains a peer and sets up a facade (a thin proxy for a peer) to receive
+// connection status notifications from the dialer and track pending request
+// counts.
+//
+// add does not add the peer to the list of peers available for choosing (the
+// Implementation).
+// The facade is responsible for adding and removing the peer from the
+// collection of available peers based on connection status notifications.
+//
+// add must be run inside a list lock.
+func (pl *List) add(id peer.Identifier) error {
+	addr := id.Identifier()
+
+	if _, ok := pl.peers[addr]; ok {
+		return peer.ErrPeerAddAlreadyInList(addr)
+	}
+
+	f := newPeerFacade(pl, id)
+	f.boundOnFinish = f.onFinish
+
+	// The transport must not call back before returning.
+	p, err := pl.transport.RetainPeer(id, f)
```

Comment:
> Yep. This may be the crux of our deadlock issues.
> 
> In #1660, I eliminated this problem for HTTP by ensuring that all notifications are synchronized and emitted from the peer status monitor goroutine. I’ll be doing the same for the gRPC and TChannel transports next.
> 
> At the moment, these notifications come from the abstract peer implementation and are triggered by start request, finish request, and dial. All the dialing for all three transports occurs in a goroutine and never on the retain/release stack.
> 

github decided I'm no longer allowed to comment on the previous thread, so continuing in a new thread.

Is there anything we can do to make sure that:
 - this is documented as an API contract (no status updates on the thread that calls RetainPeer. In general, it strictly must be non-blocking)
 - avoid the lock being held during the `RetainPeer`.

I think the first is probably the right thing to do, but just want to consider all options